src/Toolkit/Toolkit/UI/Controls/Legend/LegendDataSource.cs

```diff
@@ -0,0 +1,541 @@
+#else
+                    DependencyPropertyDescriptor.FromProperty(SceneView.SceneProperty, typeof(MapView)).AddValueChanged(sceneview, GeoViewDocumentChanged);
+#endif
+                }
+#endif
+                if (_geoview is MapView mv)
+                {
+                    _currentScale = mv.MapScale;
+                }
+            }
+
+            UpdateItemsSource();
+        }
+
+#if !XAMARIN && !XAMARIN_FORMS
+        private void GeoViewDocumentChanged(object sender, object e)
+        {
+            UpdateItemsSource();
+        }
+#endif
+
+        private void RefreshOnLoad(ILoadable loadable)
+        {
+            var listener = new Internal.WeakEventListener<ILoadable, object, EventArgs>(loadable)
+            {
+                OnEventAction = (instance, source, eventArgs) => RunOnUIThread(UpdateItemsSource),
+                OnDetachAction = (instance, weakEventListener) => instance.Loaded -= weakEventListener.OnEvent
+            };
+            loadable.Loaded += listener.OnEvent;
+        }
+
+        private void UpdateItemsSource()
+        {
+            _legendInfoTasks.Clear();
+            IEnumerable<Layer> layers = null;
+            _currentScale = double.NaN;
+            if (_geoview is MapView mv)
+            {
+                if (mv.Map != null && mv.Map.OperationalLayers == null)
+                {
+                    RefreshOnLoad(mv.Map);
+                }
+                else
+                {
+                    layers = mv.Map?.OperationalLayers;
+                }
+
+                _currentScale = mv.MapScale;
+            }
+            else if (_geoview is SceneView sv)
+            {
+                if (sv.Scene != null && sv.Scene.OperationalLayers == null)
+                {
+                    RefreshOnLoad(sv.Scene);
+                }
+                else
+                {
+                    layers = sv.Scene?.OperationalLayers;
+                }
+            }
+
+            if (layers is INotifyCollectionChanged incc)
+            {
+                var listener = new Internal.WeakEventListener<INotifyCollectionChanged, object, NotifyCollectionChangedEventArgs>(incc)
+                {
+                    OnEventAction = (instance, source, eventArgs) => Layers_CollectionChanged(source, eventArgs),
+                    OnDetachAction = (instance, weakEventListener) => instance.CollectionChanged -= weakEventListener.OnEvent
+                };
+                incc.CollectionChanged += listener.OnEvent;
+            }
+
+            TrackLayers(layers);
+
+            _items = BuildLegendList(_reverseLayerOrder ? layers?.Reverse() : layers) ?? new List<object>();
+            OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
+        }
+
+        private IEnumerable<Layer> _currentLayers;
+
+        private void TrackLayers(IEnumerable<Layer> layers)
```

Comment:
This should probably be called on GroupLayer.ChildLayers -- maybe split off GroupLayer support into a follow-up issue?